File: apps/afria-marketing-team/backend/visibility.py
from typing import Literal
from pydantic import BaseModel, Field

PARENT_ASSET_ID = "PRD-MKT-TEAM-001"
CAPABILITY = "AfrIA AI Visibility Intelligence™"
METRIC = "African Enterprise Visibility Gap™"
MISSING_DATA_POLICY = "mark_missing_never_invent"


class VisibilityAssessment(BaseModel):
    enterprise_name: str = Field(min_length=1)
    country: str = Field(min_length=1)
    verified_identity: bool = False
    verified_source_count: int = Field(default=0, ge=0)
    website_present: bool = False
    search_presence: float | None = Field(default=None, ge=0, le=100)
    ai_presence: float | None = Field(default=None, ge=0, le=100)
    media_presence: float | None = Field(default=None, ge=0, le=100)
    professional_presence: float | None = Field(default=None, ge=0, le=100)
    marketplace_presence: float | None = Field(default=None, ge=0, le=100)
    institutional_presence: float | None = Field(default=None, ge=0, le=100)
    investor_presence: float | None = Field(default=None, ge=0, le=100)


class VisibilityResult(BaseModel):
    parent_asset_id: str = PARENT_ASSET_ID
    capability: str = CAPABILITY
    metric: str = METRIC
    new_product_created: bool = False
    enterprise_name: str
    country: str
    visibility_score: float
    visibility_gap: float
    priority: Literal["critical", "high", "medium", "low"]
    confidence: Literal["high", "medium", "low"]
    observed_dimension_ratio: float
    missing_dimensions: list[str]
    recommended_actions: list[str]
    missing_data_policy: str = MISSING_DATA_POLICY


SCORE_WEIGHTS = {
    "verified_identity": 10.0,
    "verified_source_count": 5.0,
    "website_present": 5.0,
    "search_presence": 15.0,
    "ai_presence": 20.0,
    "media_presence": 10.0,
    "professional_presence": 10.0,
    "marketplace_presence": 10.0,
    "institutional_presence": 7.5,
    "investor_presence": 7.5,
}

OBSERVATION_FIELDS = [
    "search_presence",
    "ai_presence",
    "media_presence",
    "professional_presence",
    "marketplace_presence",
    "institutional_presence",
    "investor_presence",
]


def _score(payload: VisibilityAssessment) -> float:
    score = 0.0
    score += SCORE_WEIGHTS["verified_identity"] if payload.verified_identity else 0.0
    score += min(payload.verified_source_count, 5) / 5 * SCORE_WEIGHTS["verified_source_count"]
    score += SCORE_WEIGHTS["website_present"] if payload.website_present else 0.0
    for field in OBSERVATION_FIELDS:
        value = getattr(payload, field)
        normalized = 0.0 if value is None else value / 100
        score += normalized * SCORE_WEIGHTS[field]
    return round(score, 2)


def _priority(gap: float) -> Literal["critical", "high", "medium", "low"]:
    if gap >= 70:
        return "critical"
    if gap >= 50:
        return "high"
    if gap >= 30:
        return "medium"
    return "low"


def _confidence(payload: VisibilityAssessment) -> tuple[Literal["high", "medium", "low"], float, list[str]]:
    missing = [field for field in OBSERVATION_FIELDS if getattr(payload, field) is None]
    observed = len(OBSERVATION_FIELDS) - len(missing)
    ratio = round(observed / len(OBSERVATION_FIELDS), 2)
    if ratio >= 0.8:
        level: Literal["high", "medium", "low"] = "high"
    elif ratio >= 0.5:
        level = "medium"
    else:
        level = "low"
    return level, ratio, missing


def _actions(payload: VisibilityAssessment) -> list[str]:
    actions: list[tuple[float, str]] = []
    if not payload.verified_identity or payload.verified_source_count < 3:
        actions.append((15.0, "Renforcer l’Enterprise Visibility Profile™ avec identité et sources vérifiables."))
    if not payload.website_present:
        actions.append((5.0, "Créer ou vérifier le site canonique et les données structurées de l’entreprise."))

    channel_actions = {
        "search_presence": "Renforcer la découvrabilité Search/SEO avec pages canoniques, entités et preuves sourcées.",
        "ai_presence": "Renforcer la visibilité IA/AEO-GEO-LLMO avec faits structurés, citations et contenus répondant aux prompts acheteurs.",
        "media_presence": "Construire un media kit prouvé et des angles éditoriaux sans fabriquer de réputation.",
        "professional_presence": "Renforcer la présence sur les réseaux professionnels avec dirigeants, expertise et preuves commerciales.",
        "marketplace_presence": "Publier les offres vérifiées sur les marketplaces et répertoires sectoriels pertinents.",
        "institutional_presence": "Améliorer la présence dans les registres, annuaires et sources institutionnelles vérifiables.",
        "investor_presence": "Préparer un profil investisseur sourcé et les éléments de data room nécessaires à la découvrabilité financière.",
    }
    for field, action in channel_actions.items():
        value = getattr(payload, field)
        effective = 0.0 if value is None else value
        if effective < 50:
            deficit = SCORE_WEIGHTS[field] * (1 - effective / 100)
            actions.append((deficit, action))

    actions.sort(key=lambda item: item[0], reverse=True)
    return [text for _, text in actions[:5]]


def assess_visibility(payload: VisibilityAssessment) -> VisibilityResult:
    score = _score(payload)
    gap = round(100 - score, 2)
    confidence, observed_ratio, missing = _confidence(payload)
    return VisibilityResult(
        enterprise_name=payload.enterprise_name,
        country=payload.country,
        visibility_score=score,
        visibility_gap=gap,
        priority=_priority(gap),
        confidence=confidence,
        observed_dimension_ratio=observed_ratio,
        missing_dimensions=missing,
        recommended_actions=_actions(payload),
    )

File: apps/afria-marketing-team/frontend/src/visibility.ts
export const VISIBILITY_CAPABILITY = {
  parentAssetId: "PRD-MKT-TEAM-001",
  capability: "AfrIA AI Visibility Intelligence™",
  metric: "African Enterprise Visibility Gap™",
  scoringVersion: "AEVG-1.0",
  newProductCreated: false,
  missingDataPolicy: "mark_missing_never_invent"
} as const;

export type VisibilityPriority = "critical" | "high" | "medium" | "low";
export type VisibilityConfidence = "high" | "medium" | "low";

export interface VisibilityAssessment {
  enterpriseName: string;
  country: string;
  verifiedIdentity: boolean;
  verifiedSourceCount: number;
  websitePresent: boolean;
  searchPresence?: number;
  aiPresence?: number;
  mediaPresence?: number;
  professionalPresence?: number;
  marketplacePresence?: number;
  institutionalPresence?: number;
  investorPresence?: number;
}

export interface VisibilityResult {
  parentAssetId: string;
  capability: string;
  metric: string;
  scoringVersion: string;
  newProductCreated: false;
  enterpriseName: string;
  country: string;
  visibilityScore: number;
  visibilityGap: number;
  priority: VisibilityPriority;
  confidence: VisibilityConfidence;
  observedDimensionRatio: number;
  missingDimensions: string[];
  recommendedActions: string[];
  missingDataPolicy: string;
}

type ObservationField =
  | "searchPresence"
  | "aiPresence"
  | "mediaPresence"
  | "professionalPresence"
  | "marketplacePresence"
  | "institutionalPresence"
  | "investorPresence";

const WEIGHTS: Record<ObservationField, number> & {
  verifiedIdentity: number;
  verifiedSourceCount: number;
  websitePresent: number;
} = {
  verifiedIdentity: 10,
  verifiedSourceCount: 5,
  websitePresent: 5,
  searchPresence: 15,
  aiPresence: 20,
  mediaPresence: 10,
  professionalPresence: 10,
  marketplacePresence: 10,
  institutionalPresence: 7.5,
  investorPresence: 7.5
};

const OBSERVATION_FIELDS: ObservationField[] = [
  "searchPresence",
  "aiPresence",
  "mediaPresence",
  "professionalPresence",
  "marketplacePresence",
  "institutionalPresence",
  "investorPresence"
];

const ACTIONS: Record<ObservationField, string> = {
  searchPresence: "Renforcer la découvrabilité Search/SEO avec pages canoniques, entités et preuves sourcées.",
  aiPresence: "Renforcer la visibilité IA/AEO-GEO-LLMO avec faits structurés, citations et contenus répondant aux prompts acheteurs.",
  mediaPresence: "Construire un media kit prouvé et des angles éditoriaux sans fabriquer de réputation.",
  professionalPresence: "Renforcer la présence sur les réseaux professionnels avec dirigeants, expertise et preuves commerciales.",
  marketplacePresence: "Publier les offres vérifiées sur les marketplaces et répertoires sectoriels pertinents.",
  institutionalPresence: "Améliorer la présence dans les registres, annuaires et sources institutionnelles vérifiables.",
  investorPresence: "Préparer un profil investisseur sourcé et les éléments de data room nécessaires à la découvrabilité financière."
};

function clampScore(value: number | undefined): number {
  if (value === undefined || Number.isNaN(value)) return 0;
  return Math.max(0, Math.min(100, value));
}

function round2(value: number): number {
  return Math.round(value * 100) / 100;
}

function priorityForGap(gap: number): VisibilityPriority {
  if (gap >= 70) return "critical";
  if (gap >= 50) return "high";
  if (gap >= 30) return "medium";
  return "low";
}

export function assessEnterpriseVisibility(input: VisibilityAssessment): VisibilityResult {
  let score = 0;
  score += input.verifiedIdentity ? WEIGHTS.verifiedIdentity : 0;
  score += (Math.min(Math.max(input.verifiedSourceCount, 0), 5) / 5) * WEIGHTS.verifiedSourceCount;
  score += input.websitePresent ? WEIGHTS.websitePresent : 0;

  const missingDimensions: string[] = [];
  for (const field of OBSERVATION_FIELDS) {
    const raw = input[field];
    if (raw === undefined) missingDimensions.push(field);
    score += (clampScore(raw) / 100) * WEIGHTS[field];
  }

  const visibilityScore = round2(score);
  const visibilityGap = round2(100 - visibilityScore);
  const observedDimensionRatio = round2((OBSERVATION_FIELDS.length - missingDimensions.length) / OBSERVATION_FIELDS.length);
  const confidence: VisibilityConfidence = observedDimensionRatio >= 0.8 ? "high" : observedDimensionRatio >= 0.5 ? "medium" : "low";

  const rankedActions: Array<{ deficit: number; action: string }> = [];
  if (!input.verifiedIdentity || input.verifiedSourceCount < 3) {
    rankedActions.push({ deficit: 15, action: "Renforcer l’Enterprise Visibility Profile™ avec identité et sources vérifiables." });
  }
  if (!input.websitePresent) {
    rankedActions.push({ deficit: 5, action: "Créer ou vérifier le site canonique et les données structurées de l’entreprise." });
  }
  for (const field of OBSERVATION_FIELDS) {
    const value = clampScore(input[field]);
    if (value < 50) {
      rankedActions.push({ deficit: WEIGHTS[field] * (1 - value / 100), action: ACTIONS[field] });
    }
  }
  rankedActions.sort((a, b) => b.deficit - a.deficit);

  return {
    ...VISIBILITY_CAPABILITY,
    enterpriseName: input.enterpriseName,
    country: input.country,
    visibilityScore,
    visibilityGap,
    priority: priorityForGap(visibilityGap),
    confidence,
    observedDimensionRatio,
    missingDimensions,
    recommendedActions: rankedActions.slice(0, 5).map(item => item.action)
  };
}
